File: 0386_Lexicographical_Numbers/Solution.java
import java.util.ArrayList;

public class Solution {
    public List<Integer> lexicalOrder(int n) {
        ArrayList<Integer> answer = new ArrayList<>();
        // roots
        for (int i = 1; i < 10; ++i) {
        	dfs(i, n, answer);
        }
        return answer;
    }

    dfs(int root, int n, List<Integer> results) {
    	if (root > n) {
    		return;
    	} else {
    		// visit current node
    		results.add(root)
    		// visit all child nodes
    		// only 0 - 9 anyway
    		int nextLevelRoot = 10 * root;
    		for (int i = 0; i < 10; ++i) {
    			if (nextLevelRoot + i > n) {
    				return;
    			}
    			dfs(nextLevelRoot + i, n, results);
    		}
    	}
    }
}

File: 0386_Lexicographical_Numbers/Solution.py
class Solution(object):
    def lexicalOrder(self, n):
        """
        :type n: int
        :rtype: List[int]
        """

        answer = list()
        for i in range(1, 10):
            if i > n:
                return answer
            self.dfs(i, n, answer)
        return answer

    def dfs(self, root, n, answer):
        # visit node
        answer.append(root)
        # visit all child node
        # only 0 - 9 ten digits anyway
        next_level_node = root * 10
        for i in range(next_level_node, next_level_node + 10):
            if i > n:
                return
            self.dfs(i, n, answer)


print Solution().lexicalOrder(49999)
